source/vir/src/func_to_air.rs

```diff
@@ -119,6 +126,24 @@ fn func_body_to_air(
         decl_commands.push(Arc::new(CommandX::Global(fuel_axiom)));
     }
 
+    // For trait method implementations, use trait method function name and add Self type argument
+    let mut typ_args = vec_map(&function.x.typ_params(), |x| Arc::new(TypX::TypParam(x.clone())));
+    let name = if let FunctionKind::TraitMethodImpl {
+        method,
+        trait_typ_args,
+        datatype,
+        datatype_typ_args,
+        ..
+    } = &function.x.kind
+    {
+        let self_typ = Arc::new(TypX::Datatype(datatype.clone(), datatype_typ_args.clone()));
+        typ_args = vec![self_typ];
+        typ_args.append(&mut (**trait_typ_args).clone());
```

Comment:
> If we wanted to support something like this, we'd need to generate an SMT encoding that dispatches not just on Self, but on other type parameters like the A in trait T<A>.

Are you not already doing this? I had thought that was what this was for:

```
typ_args.append(&mut (**trait_typ_args).clone());
```

Reply:
For the second question, this might be an example of what you're talking about:

```
trait T {
    #[spec]
    fn f(&self) -> Box<Self> { no_method_body() }

    #[spec]
    fn g(&self) -> Box<Self> { no_method_body() }

    #[proof]
    fn p(&self) {
        ensures(!equal(self.f(), self.g()));
        no_method_body()
    }
}
#[proof]
#[verifier(broadcast_forall)]
fn f_not_g<A: T>(a: &A) {
    ensures(!equal(a.f(), a.g()));
    a.p();
}
```

Here, we can't implement trait `T` for a type `struct S {}`, because there's only one value of type `S`, so we can't have `f() != g()`.  So it would be important that we don't try to encode `f_not_g` as saying for all types there are two values of that type -- it needs to be limited to types that satisfy trait `T`.  (It's also important that we don't assume `S: T` while we're trying to prove `S: T`, since this would be circular reasoning.)

We don't support `broadcast_forall` for `A: T` yet, but this is something to keep in mind when we do.  We'll probably need an SMT predicate to represent `A: T`.